# Single Cell Sequencing Quality Assessment: SCQUA

SCQUA is a program to help assessing the sequencing quality of single-cell data. It measures the sensitivity (detection limit) and the accuracy (correlation) according to the spike-ins (reference RNA).

It has performed analysis for a couple of single-cell sequencing benchmarking papers:

[Comparative analysis of sequencing technologies for single-cell transcriptomics, KN Natarajan, Z Miao, et al. Genome Biology volume 20, Article number: 70 (2019) ](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1676-5)

[High-throughput Full-Length Single-Cell RNA-Seq Automation L Mamanova, Z Miao, et al. (2020)]()

All the codes processing these data and making the figures are available in the [example](https://github.com/chichaumiau/SCQUA/tree/master/examples) folder.

## installation
To install the program please use the codes below:

> git clone https://github.com/chichaumiau/SCQUA.git
> cd SCQUA
> python setup.py install

## usage
Please refer to the [examples](https://github.com/chichaumiau/SCQUA/tree/master/examples) for details about how to use the program. 


## Reference
Reference to the paper originated the measurement of scRNA-seq sensitivity and accuracy:

[Power analysis of single-cell RNA-sequencing experiments
V Svensson, KN Natarajan, et al. Nature Methods volume 14, pages381â€“387(2017)](https://www.nature.com/articles/nmeth.4220)
